Catappult/appcoins-wallet-android
# [feature/WAL-57] In-App Carrier Billing

**What does this PR do?**

   This PR implements carrier billing through dimoco/onebip for in-app purchases 📞 💸 . It should follow the standards that we defined recently, so don't hold back on criticisms (I probably missed something). 

**Database changed?**

   No

**Where should the reviewer start?**

- Most of the code is located in com.asfoundation.wallet.ui.iab.payments and com.appcoins.wallet.billing.carrierbilling

**How should this be manually tested?**

  You need to tests different flows and different configurations to make sure everything is ok, this includes:
 - Success flow, i.e. make an in-app purchase succesfully through carrier billing. Note that this only works with certain carriers in certain countries (for now DE, Vodafone). You can ask me or André for more info about this.
 - Error flows, this can include inputting an invalid phone number or an unsupported carrier or country. This can also include lack of network, blocked wallet, etc. Also, an important one to check is exceeding the value limit for the carrier, i.e. some carriers only allow purchases up to a certain value (e.g. 23.59€) and above a certain value (e.g. 1€).
 - Check if layouts are ok both in portrait and landscape modes. Try to change orientation in different circumstances (e.g. rotating while we're validating a phone number, etc.).
 - Validate analytics and make sure they're being sent (you'll probably need to talk with me to validate all cases)
 - Check canceling flows, such as pressing system back in webview, or in the "fee screen" and see if it works as expected

**What are the relevant tickets?**

  [WAL-57](https://aptoide.atlassian.net/browse/WAL-57)

**Questions:**

   It adds two new dependencies: WorkManager and core-ktx. Although I didn't end up using, WorkManager is important to do deferred work and we will probably use it in the future. Core-ktx is simply a collection of useful Kotlin extensions for the Android framework/libraries. I use it here to simplify spanned strings.


**Code Review Checklist**
